Added a delegate to Scaffold triggered when removing the bottom insets

matheus-kirchesch-old opened the pull request:
What is solves: Gives access to the removed bottom inset to any child widget without them having to disable the super useful scaffold remove bottom insets functionality, the reason this bottom inset is useful is that you may want to do your own animations based on the bottom inset, you may want to use it in a widget that overflows the main widget to control its position, there are tons of uses for this info, i am using it to control a persistent bottom "header" of my pages, it behaves like a scaffold bottomSheet, but i don't need a scaffold for it, it is instantiated way before any page specific scaffolds

Bellow is the part before the edit
---
Reason: The removed bottom insets is very useful for me at least, I use it to manipulate stuff that "hovers" over main screens, etc...

matheus-kirchesch committed:
Added a delegate to Scaffold triggered when removing the bottom insets

Reason: The removed bottom insets is very useful for me at least, I use it to manipulate stuff that "hovers" over main screens, etc...

matheus-kirchesch committed:
View insets removal delegate for scaffold tests added

matheus-kirchesch committed:
Trailling line hotfix

goderbauer commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -2581,6 +2581,9 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin, Resto
     super.dispose();
   }
 
+  /// Delegate triggered when we remove the bottom insets
```
avoid "we" in doc comments, see style guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-the-passive-voice-recommend-do-not-require

goderbauer replied:
Doc comments should be grammatically correct sentences and end with a `.`.

goderbauer replied:
The doc comments needs to explain in a little more detail when this fires and why someone should care.

goderbauer commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -2581,6 +2581,9 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin, Resto
   }
 
+  /// Delegate triggered when we remove the bottom insets
+  void Function(EdgeInsets edgeInsets)? onRemoveBottomInsets;
```
please use a typedef.

goderbauer replied:
also, this should really be a configuration option on the Scaffold, not a random property that anybody can set on the state.

goderbauer commented on packages/flutter/lib/src/material/scaffold.dart:
```diff
@@ -2599,6 +2602,7 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin, Resto
       removeBottom: removeBottomPadding,
     );
     if (removeBottomInset) {
+      onRemoveBottomInsets?.call(data.viewInsets);
```
Calling a callback from a build function is an odd side effect of build. Also, whether the bottomInsets are removed is user-controlled via the `resizeToAvoidBottomInset` property. Why do you need a callback for something that's fully in your control?

Kirchesch replied:
Because resizeToAvoidBottomInset is very useful, and sometimes you want to use the bottom inset to control something that is outside the scaffold, this just keeps track of the removed bottom inset

goderbauer commented on packages/flutter/test/material/scaffold_test.dart:
```diff
@@ -2621,6 +2621,41 @@ void main() {
+    await tester.showKeyboard(textFieldFinder);
+    await tester.pump(const Duration(milliseconds: 400));
+
+    // I don't know how to make the keyboard really open here and add the view
+    //  insets, i have tested this in my project and it works, its a simple
+    //  feature that doesn't interact with anything else
+    // expect(removedBottomInset != null, true);
```
You can use a custom MediaQuery in the test that simulates having a bottomInset.

goderbauer reviewed (commented):
It is not clear to me what problem this is solving. I would suggest filing an issue that describes the problem first before suggesting a solution.

matheus-kirchesch committed:
PR fixes

- Moved the delegate to the widget itself
- Rewritten the comments
- Using typedef

matheus-kirchesch-old committed:
Merge branch 'flutter:master' into master

goderbauer commented:
I am going to close this as this isn't really adding any functionality. As explained in https://github.com/flutter/flutter/pull/107669#discussion_r922560416, wether the bottomInsets are removed is user-controlled via the resizeToAvoidBottomInset property and having a callback doesn't appear to make much sense.
